### PR TITLE
[url_launcher] Added draggable property [#102735]

### PR DESCRIPTION
**Pull request by kristoffer-zliide**
Allowing use of the Link widget in lists with mouse drag enabled.

In a desktop web browser, links are be default draggable. If they are put in e.g. a list, then that list cannot be scrolled by dragging it with the mouse. This PR makes it possible to disable this behavior.

Issues fixed: [#102735](https://github.com/flutter/flutter/issues/102735)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
@ditman Ping on this review?

**Comment by ditman** [comment]
Apologies! I'm going to bump this to @mdebbar, he knows the `Link` widget much better than I do.

Mouad, do you mind taking a look? Thanks!!

**Comment by stuartmorgan** [comment]
From triage: @mdebbar Ping on this review.

File: packages/url_launcher/url_launcher_web/example/pubspec.yaml
Comment (mdebbar) [inline packages/url_launcher/url_launcher_web/example/pubspec.yaml:17]: @ditman I'm not qualified to review this part of the PR. Could you take a look?

**Review by mdebbar — APPROVED** [review]
The PR looks good to me. I would like for @ditman to take a look at the pubspec changes since I'm not very familiar with how plugin dependencies are supposed to be.

File: packages/url_launcher/url_launcher_web/lib/src/link.dart
Comment (ditman) [inline packages/url_launcher/url_launcher_web/lib/src/link.dart:259]: I was checking caniuse.com, and it seems that:

* there's no unprefixed `user-drag` property. [Docs](https://caniuse.com/webkit-user-drag).
* wouldn't `draggable = false` be enough?

Also, if disabling drag in the element, is it necessary to stop the default drag event?

I think this method could be much simplified to be just `_element.draggable = false`?

File: packages/url_launcher/url_launcher_web/example/pubspec.yaml
Comment (ditman) [reply packages/url_launcher/url_launcher_web/example/pubspec.yaml:17]: This will need to be reverted so we can merge and publish the affected packages in the correct order.

**Review by ditman — APPROVED** [review]
We cannot merge this PR as is, because we now need to update version and CHANGELOG of the affected packages. Could you prepare a separate PR for each of the packages, so we can merge and publish them in order?

**Comment by stuartmorgan** [comment]
> Could you prepare a separate PR for each of the packages, so we can merge and publish them in order?

We actually just do this one at a time; see https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-federated-plugins

So the next step is to make just the platform interface PR, then once that lands, the implemenation package PR, then finally rebase this one.

File: packages/url_launcher/url_launcher_web/lib/src/link.dart
Comment (kristoffer-zliide) [reply packages/url_launcher/url_launcher_web/lib/src/link.dart:259]: It seems you're close to being right. We started out with the styling approach since it allowed us to apply a global work-around in index.html (`<style>a { -webkit-user-drag: none; }</style>`). Probably just added `user-drag` since it was right there in `html_dart2js.dart`, and if it works, it's nicer than the `-webkit-` style. But it also seems that `draggable` works on both Chrome and Edge; however, I couldn't make it work on Firefox without `preventDefault`. In the end, I guess it's a matter of taste: Do you want style or properties for this? Do you want to wrap the drag subscription in some Firefox shim?

Comment (ditman) [reply packages/url_launcher/url_launcher_web/lib/src/link.dart:259]: After some tests, it seems that **the only solution that seems to work the same across all platforms is `preventDefault` on the `dragstart` event.** See:

* https://jsfiddle.net/ditman/2kLuo8vd

If `draggable=false`, none of the browsers will trigger a `dragstart` event (as expected).

In Firefox, `draggable=false` works (it won't let you drag the Link), **but** it'll also start selecting from the text of the link as the user drags (which I guess is what you couldn't get to work?). This can be "corrected" with `user-select`, but then the Link won't be highlighted at all when text is selected from the outside. This looks like a Firefox bug/feature, but I couldn't find anything relevant in [their bug tracker](https://bugzilla.mozilla.org/buglist.cgi?bug_status=__open__&component=DOM%3A%20Selection&product=Core&query_format=advanced&order=bug_status%2Cpriority%2Cassigned_to%2Cbug_id&limit=0).

Comment (ditman) [reply packages/url_launcher/url_launcher_web/lib/src/link.dart:259]: (My second favorite version might be `draggable=false` and if the browser is firefox, prevent the `selectstart` event (`draggable=false + no selectstart` in the fiddle), which doesn't interfere with the selection from the outside as the `user-select` CSS attribute does!)

Comment (kristoffer-zliide) [reply packages/url_launcher/url_launcher_web/lib/src/link.dart:259]: Actually, what we couldn't get to work is [the issue I linked to](https://github.com/flutter/flutter/issues/102735): links in a list. For that case preventing select start doesn't seem to be sufficient. But interactions with text selection should indeed probably also be considered.

**Comment by kristoffer-zliide** [comment]
@stuartmorgan: Yeah, but I'm not sure if we've reached a conclusion on this: https://github.com/flutter/plugins/pull/5430/files#r896078916. I can make pull requests that adds draggable and preventDefault (ie, what's done for Firefox here: https://github.com/flutter/plugins/pull/5430/files#diff-77a04b3b00b00a4480b7d5e75e089dde7cca99ccebcc4f43e0e25d2bba5346e0R255), then you can always tweak it to your heart's content afterwards?

**Comment by stuartmorgan** [comment]
@ditman You marked this as approved; what was your take on the resolution of that comment thread?

**Comment by stuartmorgan** [comment]
@kristoffer-zliide It sounds from the above like this was ready for a platform interface PR; are you still planning on splitting that out to move this forward?

**Comment by kristoffer-zliide** [comment]
@stuartmorgan: [here](https://github.com/flutter/plugins/pull/6282) you go

**Comment by stuartmorgan** [comment]
Status update from triage: blocked on resolving the breaking change issue in https://github.com/flutter/plugins/pull/6282.

**Comment by stuartmorgan** [comment]
Status from triage: still blocked on https://github.com/flutter/plugins/pull/6282

**Comment by stuartmorgan** [comment]
Since this is still blocked on the other PR, I'm going to go ahead and mark this as a draft just to avoid having it come up in our triage queue. Once the blocking issue is resolved, this can be marked as ready for review again.

**Comment by stuartmorgan** [comment]
Closing since https://github.com/flutter/plugins/pull/6282 is now closed.

**Pull request closed without merging**
